### PR TITLE
CLOUDSTACK-8751 minimise downtime of network when network update is in progress.

### PR DESCRIPTION
**Pull request by bvbharatk**
more info in the bug link below.
https://issues.apache.org/jira/browse/CLOUDSTACK-8751

### CODE REVIEW

**Comment by insom** [comment]
I really like the look of this feature. I don't want to come across as too negative with respect to some niggles with the code formatting.

**Comment by remibergsma** [comment]
@bvbharatk Thanks! Can you please rebase against current master? If you do, we can run the tests.

Did you run the BVT tests against this branch? Please post output.

**Comment by bhaisaab** [comment]
Any update on this @bvbharatk ?

**Comment by bvbharatk** [comment]
Hi, 

rebased with master.

**Comment by remibergsma** [comment]
LGTM, based on a set of tests that I run on this branch (which I rebased myself first):

```
nosetests --with-marvin --marvin-config=${marvinCfg} -s -a tags=advanced,required_hardware=true \
component/test_vpc_redundant.py \
component/test_routers_iptables_default_policy.py \
component/test_routers_network_ops.py \
component/test_vpc_router_nics.py \
smoke/test_loadbalance.py \
smoke/test_internal_lb.py \
smoke/test_ssvm.py \
smoke/test_network.py

```

Result:

```
Create a redundant VPC with two networks with two VMs in each network ... === TestName: test_01_create_redundant_VPC_2tiers_4VMs_4IPs_4PF_ACL | Status : SUCCESS ===
ok
Create a redundant VPC with two networks with two VMs in each network and check default routes ... === TestName: test_02_redundant_VPC_default_routes | Status : SUCCESS ===
ok
Test iptables default INPUT/FORWARD policy on RouterVM ... === TestName: test_02_routervm_iptables_policies | Status : SUCCESS ===
ok
Test iptables default INPUT/FORWARD policies on VPC router ... === TestName: test_01_single_VPC_iptables_policies | Status : SUCCESS ===
ok
Stop existing router, add a PF rule and check we can access the VM ... === TestName: test_isolate_network_FW_PF_default_routes | Status : SUCCESS ===
ok
Test redundant router internals ... === TestName: test_RVR_Network_FW_PF_SSH_default_routes | Status : SUCCESS ===
ok
Create a VPC with two networks with one VM in each network and test nics after destroy ... === TestName: test_01_VPC_nics_after_destroy | Status : SUCCESS ===
ok
Create a VPC with two networks with one VM in each network and test default routes ... === TestName: test_02_VPC_default_routes | Status : SUCCESS ===
ok
Check the password file in the Router VM ... === TestName: test_isolate_network_password_server | Status : SUCCESS ===
ok
Check that the /etc/dhcphosts.txt doesn't contain duplicate IPs ... === TestName: test_router_dhcphosts | Status : SUCCESS ===
ok
Test to create Load balancing rule with source NAT ... === TestName: test_01_create_lb_rule_src_nat | Status : SUCCESS ===
ok
Test to create Load balancing rule with non source NAT ... === TestName: test_02_create_lb_rule_non_nat | Status : SUCCESS ===
ok
Test for assign & removing load balancing rule ... === TestName: test_assign_and_removal_lb | Status : SUCCESS ===
ok
Test to verify access to loadbalancer haproxy admin stats page ... === TestName: test02_internallb_haproxy_stats_on_all_interfaces | Status : SUCCESS ===
ok
Test create, assign, remove of an Internal LB with roundrobin http traffic to 3 vm's ... === TestName: test_01_internallb_roundrobin_1VPC_3VM_HTTP_port80 | Status : SUCCESS ===
ok
Test SSVM Internals ... === TestName: test_03_ssvm_internals | Status : SUCCESS ===
ok
Test CPVM Internals ... === TestName: test_04_cpvm_internals | Status : SUCCESS ===
ok
Test stop SSVM ... === TestName: test_05_stop_ssvm | Status : SUCCESS ===
ok
Test stop CPVM ... === TestName: test_06_stop_cpvm | Status : SUCCESS ===
ok
Test reboot SSVM ... === TestName: test_07_reboot_ssvm | Status : SUCCESS ===
ok
Test reboot CPVM ... === TestName: test_08_reboot_cpvm | Status : SUCCESS ===
ok
Test destroy SSVM ... === TestName: test_09_destroy_ssvm | Status : SUCCESS ===
ok
Test destroy CPVM ... === TestName: test_10_destroy_cpvm | Status : SUCCESS ===
ok
Test for port forwarding on source NAT ... === TestName: test_01_port_fwd_on_src_nat | Status : SUCCESS ===
ok
Test for port forwarding on non source NAT ... === TestName: test_02_port_fwd_on_non_src_nat | Status : SUCCESS ===
ok
Test for reboot router ... === TestName: test_reboot_router | Status : SUCCESS ===
ok
Test for Router rules for network rules on acquired public IP ... === TestName: test_network_rules_acquired_public_ip_1_static_nat_rule | Status : SUCCESS ===
ok
Test for Router rules for network rules on acquired public IP ... === TestName: test_network_rules_acquired_public_ip_2_nat_rule | Status : SUCCESS ===
ok
Test for Router rules for network rules on acquired public IP ... === TestName: test_network_rules_acquired_public_ip_3_Load_Balancer_Rule | Status : SUCCESS ===
ok

----------------------------------------------------------------------
Ran 29 tests in 12815.696s

OK
```


And:

```
nosetests --with-marvin --marvin-config=${marvinCfg} -s -a tags=advanced,required_hardware=false \
smoke/test_routers.py \
smoke/test_network_acl.py \
smoke/test_privategw_acl.py \
smoke/test_reset_vm_on_reboot.py \
smoke/test_vm_life_cycle.py \
smoke/test_vpc_vpn.py \
smoke/test_service_offerings.py \
component/test_vpc_offerings.py \
component/test_vpc_routers.py
```

Result:

```
Test router internal advanced zone ... === TestName: test_02_router_internal_adv | Status : SUCCESS ===
ok
Test restart network ... === TestName: test_03_restart_network_cleanup | Status : SUCCESS ===
ok
Test router basic setup ... === TestName: test_05_router_basic | Status : SUCCESS ===
ok
Test router advanced setup ... === TestName: test_06_router_advanced | Status : SUCCESS ===
ok
Test stop router ... === TestName: test_07_stop_router | Status : SUCCESS ===
ok
Test start router ... === TestName: test_08_start_router | Status : SUCCESS ===
ok
Test reboot router ... === TestName: test_09_reboot_router | Status : SUCCESS ===
ok
test_privategw_acl (integration.smoke.test_privategw_acl.TestPrivateGwACL) ... === TestName: test_privategw_acl | Status : SUCCESS ===
ok
Test reset virtual machine on reboot ... === TestName: test_01_reset_vm_on_reboot | Status : SUCCESS ===
ok
Test advanced zone virtual router ... === TestName: test_advZoneVirtualRouter | Status : SUCCESS ===
ok
Test Deploy Virtual Machine ... === TestName: test_deploy_vm | Status : SUCCESS ===
ok
Test Multiple Deploy Virtual Machine ... === TestName: test_deploy_vm_multiple | Status : SUCCESS ===
ok
Test Stop Virtual Machine ... === TestName: test_01_stop_vm | Status : SUCCESS ===
ok
Test Start Virtual Machine ... === TestName: test_02_start_vm | Status : SUCCESS ===
ok
Test Reboot Virtual Machine ... === TestName: test_03_reboot_vm | Status : SUCCESS ===
ok
Test destroy Virtual Machine ... === TestName: test_06_destroy_vm | Status : SUCCESS ===
ok
Test recover Virtual Machine ... === TestName: test_07_restore_vm | Status : SUCCESS ===
ok
Test migrate VM ... === TestName: test_08_migrate_vm | Status : SUCCESS ===
ok
Test destroy(expunge) Virtual Machine ... === TestName: test_09_expunge_vm | Status : SUCCESS ===
ok
Test Remote Access VPN in VPC ... === TestName: test_vpc_remote_access_vpn | Status : SUCCESS ===
ok
Test VPN in VPC ... === TestName: test_vpc_site2site_vpn | Status : SUCCESS ===
ok
Test to create service offering ... === TestName: test_01_create_service_offering | Status : SUCCESS ===
ok
Test to update existing service offering ... === TestName: test_02_edit_service_offering | Status : SUCCESS ===
ok
Test to delete service offering ... === TestName: test_03_delete_service_offering | Status : SUCCESS ===
ok
Test for delete account ... === TestName: test_delete_account | Status : SUCCESS ===
ok
Test for Associate/Disassociate public IP address for admin account ... === TestName: test_public_ip_admin_account | Status : SUCCESS ===
ok
Test for Associate/Disassociate public IP address for user account ... === TestName: test_public_ip_user_account | Status : SUCCESS ===
ok
Test for release public IP address ... === TestName: test_releaseIP | Status : SUCCESS ===
ok
Test create VPC offering ... === TestName: test_01_create_vpc_offering | Status : SUCCESS ===
ok
Test VPC offering without load balancing service ... === TestName: test_03_vpc_off_without_lb | Status : SUCCESS ===
ok
Test VPC offering without static NAT service ... === TestName: test_04_vpc_off_without_static_nat | Status : SUCCESS ===
ok
Test VPC offering without port forwarding service ... === TestName: test_05_vpc_off_without_pf | Status : SUCCESS ===
ok
Test VPC offering with invalid services ... === TestName: test_06_vpc_off_invalid_services | Status : SUCCESS ===
ok
Test update VPC offering ... === TestName: test_07_update_vpc_off | Status : SUCCESS ===
ok
Test list VPC offering ... === TestName: test_08_list_vpc_off | Status : SUCCESS ===
ok
Test start/stop of router after addition of one guest network ... === TestName: test_01_start_stop_router_after_addition_of_one_guest_network | Status : SUCCESS ===
ok
Test reboot of router after addition of one guest network ... === TestName: test_02_reboot_router_after_addition_of_one_guest_network | Status : SUCCESS ===
ok
Test to change service offering of router after addition of one guest network ... === TestName: test_04_chg_srv_off_router_after_addition_of_one_guest_network | Status : SUCCESS ===
ok
Test destroy of router after addition of one guest network ... === TestName: test_05_destroy_router_after_addition_of_one_guest_network | Status : SUCCESS ===
ok
Test to stop and start router after creation of VPC ... === TestName: test_01_stop_start_router_after_creating_vpc | Status : SUCCESS ===
ok
Test to reboot the router after creating a VPC ... === TestName: test_02_reboot_router_after_creating_vpc | Status : SUCCESS ===
ok
Tests to change service offering of the Router after ... === TestName: test_04_change_service_offerring_vpc | Status : SUCCESS ===
ok
Test to destroy the router after creating a VPC ... === TestName: test_05_destroy_router_after_creating_vpc | Status : SUCCESS ===
ok

----------------------------------------------------------------------
Ran 44 tests in 9022.348s

OK
```

These test do not cover your change, all they do is show you didn't break them. Someone else needs to review the code.

**Comment by remibergsma** [comment]
Pinging @wilderrodrigues to have a look.

@bvbharatk Is there any (integration) test to test this feature?

File: server/test/com/cloud/network/element/VirtualRouterElementTest.java [hunk 1]
```diff
@@ -225,6 +235,62 @@ public void testPrepare() {
 
     }
 
+    @Test
+    public void testGetRouters1(){
```
Comment (DaanHoogland) [inline server/test/com/cloud/network/element/VirtualRouterElementTest.java]: testGetRoutersUnderThisOrTheseCircumstances() ould be a more appropriate name.
testGetRouters1, testGetRouters2 and testGetRouters3 is not really clear.

**Comment by DaanHoogland** [comment]
good that you found the time to implement this feature @bvbharatk , thanks. an integrations test for this is needed as we will judge future PRs against it. As in, they may not break this. I have no problem with the code if it can be shown to do what it is supposed to. If I find time I will help to prove it but that won't be the coming weeks.

**Comment by bvbharatk** [comment]
@DaanHoogland 

Added a test to verify if the routers are getting updated one after the other.

Test create network with redundant routers ... === TestName: test_createRvRNetwork | Status : SUCCESS === 
ok
Test create network with non-default guest cidr with redundant routers ... === TestName: test_createRvRNetwork | Status : SUCCESS === 
ok
Test create RvR supported network offering ... === TestName: test_createRvRNetworkOffering | Status : SUCCESS === 
ok
Test redundant router internals ... === TestName: test_redundantVR_internals | Status : SUCCESS === 
ok
Test stop master RVR ... === TestName: test_01_stopMasterRvR | Status : SUCCESS === 
ok
Test stop backup RVR ... === TestName: test_02_stopBackupRvR | Status : SUCCESS === 
ok
Test reboot master RVR ... === TestName: test_03_rebootMasterRvR | Status : SUCCESS === 
ok
Test reboot backup RVR ... === TestName: test_04_rebootBackupRvR | Status : SUCCESS === 
ok
Test stop backup RVR and start instance ... === TestName: test_05_stopBackupRvR_startInstance | Status : SUCCESS === 
ok
Test update network and check if VRs are updated in sequence ... === TestName: test_06_updateVRs_in_sequence | Status : SUCCESS === 
ok

----------------------------------------------------------------------
Ran 10 tests in 5941.052s

OK

**Comment by ustcweizhou** [comment]
good, this might also be useful for restartnetwork, see #1198

**Comment by bhaisaab** [comment]
@bvbharat please squash the changes into a single commit

ping, re-review please

**Comment by deepthimachiraju** [comment]
Tested this feature Manually . 
LGTM .

**Comment by ustcweizhou** [comment]
@deepthimachiraju what's the downtime of network ?
I think this PR is too complicated.

**Comment by rhtyd** [comment]
@bvbharatk please rebase against latest master and push -f, update on status of your PR

please move the sql changes to 481-490 upgrade path

tag:easypr

**Comment by bvbharatk** [comment]
rebased with master and moved the db changes to schema481to490

**Comment by bvbharatk** [comment]
### ACS CI BVT Run
 **Sumarry:**
 Build Number 32
 Hypervisor xenserver
 NetworkType Advanced
 Passed=184
 Failed=4
 Skipped=6

**Failed tests:**
* test_guest_vlan_range.py

 * ContextSuite context=TestDedicateGuestVlanRange>:teardown Failing since 10 runs

* test_list_ids_parameter.py

 * ContextSuite context=TestListIdsParams>:setup Failed


**Skipped tests:**
test_vm_nic_adapter_vmxnet3
test_static_role_account_acls
test_deploy_vgpu_enabled_vm
test_vm_nic_adapter_vmxnet3
test_static_role_account_acls
test_deploy_vgpu_enabled_vm

**Passed test suits:**
test_deploy_vm_with_userdata.py
test_affinity_groups_projects.py
test_portable_publicip.py
test_vpc_vpn.py
test_over_provisioning.py
test_global_settings.py
test_scale_vm.py
test_service_offerings.py
test_routers_iptables_default_policy.py
test_routers.py
test_reset_vm_on_reboot.py
test_snapshots.py
test_deploy_vms_with_varied_deploymentplanners.py
test_network.py
test_non_contigiousvlan.py
test_deploy_vm_iso.py
test_public_ip_range.py
test_multipleips_per_nic.py
test_regions.py
test_affinity_groups.py
test_network_acl.py
test_pvlan.py
test_volumes.py
test_nic.py
test_deploy_vm_root_resize.py
test_resource_detail.py
test_secondary_storage.py
test_dynamicroles.py
test_disk_offerings.py
test_deploy_vm_with_userdata.py
test_affinity_groups_projects.py
test_portable_publicip.py
test_vpc_vpn.py
test_over_provisioning.py
test_global_settings.py
test_scale_vm.py
test_service_offerings.py
test_routers_iptables_default_policy.py
test_routers.py
test_reset_vm_on_reboot.py
test_snapshots.py
test_deploy_vms_with_varied_deploymentplanners.py
test_network.py
test_non_contigiousvlan.py
test_deploy_vm_iso.py
test_public_ip_range.py
test_multipleips_per_nic.py
test_regions.py
test_affinity_groups.py
test_network_acl.py
test_pvlan.py
test_volumes.py
test_nic.py
test_deploy_vm_root_resize.py
test_resource_detail.py
test_secondary_storage.py
test_dynamicroles.py
test_disk_offerings.py

**Comment by bvbharatk** [comment]
Hi,

In the CI the test have failed because of a tear down issue i.e. the test failed to clean the used resources. These failure can be ignored until we fix the test.

**Comment by swill** [comment]
@bvbharatk this PR has merge conflicts.  Can you rebase and fix the merge conflicts and re-push?  Thanks...

**Comment by bvbharatk** [comment]
@swill 
Rebased with master and resolved the conflicts.

**Comment by swill** [comment]
@bvbharatk can you close and reopen this PR to kick off jenkins again to see if we can get this green?

I will rerun CI on it to make sure we are still good after the fixes to the merge conflicts.  Thanks...

**Comment by swill** [comment]
I have attempted to run CI against this one 3 times in a row on different CI environments.  Every time it is failing with the following:

```
==== Deploy DC Started ====
Exception Occurred :['Traceback (most recent call last):\n', '  File "/data/git/cs1/cloudstack/tools/marvin/marvin/deployDataCenter.py", line 142, in addHosts\n    ret = self.__apiClient.addHost(hostcmd)\n', '  File "/usr/lib/python2.7/site-packages/marvin/cloudstackAPI/cloudstackAPIClient.py", line 1532, in addHost\n    response = self.connection.marvinRequest(command, response_type=response, method=method)\n', '  File "/usr/lib/python2.7/site-packages/marvin/cloudstackConnection.py", line 379, in marvinRequest\n    raise e\n', 'CloudstackAPIException: Execute cmd: addhost failed, due to: errorCode: 530, errorText:Unable to add the host\n']
Exception Occurred :['Traceback (most recent call last):\n', '  File "/data/git/cs1/cloudstack/tools/marvin/marvin/deployDataCenter.py", line 142, in addHosts\n    ret = self.__apiClient.addHost(hostcmd)\n', '  File "/usr/lib/python2.7/site-packages/marvin/cloudstackAPI/cloudstackAPIClient.py", line 1532, in addHost\n    response = self.connection.marvinRequest(command, response_type=response, method=method)\n', '  File "/usr/lib/python2.7/site-packages/marvin/cloudstackConnection.py", line 379, in marvinRequest\n    raise e\n', 'CloudstackAPIException: Execute cmd: addhost failed, due to: errorCode: 530, errorText:Guid is not updated for cluster with specified cluster id; need to wait for hosts in this cluster to come up\n']

===deploy dc failed, so cleaning the created entries===

====DeployDC: CleanUp Started====

====Clean Up Entries=== {'PhysicalNetwork': [u'c2fbba7e-cb36-4eab-9441-b66b16bdda01'], 'Cluster': [u'08ff440f-806f-43e2-a51e-9a54b7364a1a'], 'Pod': [u'f32f353d-5ba8-4447-a71f-dbb3eeec6296'], 'order': ['Cluster', 'Pod', 'PhysicalNetwork', 'Zone'], 'Zone': [u'085c8908-9cb8-47d2-89b3-0b53b8b882ce']}

===Removing DataCenter Failed===
Thu May 26 18:43:19 CEST 2016
Deployment failed, please investigate!
```

**Comment by bvbharatk** [comment]
### ACS CI BVT Run
 **Sumarry:**
 Build Number 109
 Hypervisor xenserver
 NetworkType Advanced
 Passed=72
 Failed=1
 Skipped=3

_Link to logs Folder (search by build_no):_ https://www.dropbox.com/sh/yj3wnzbceo9uef2/AAB6u-Iap-xztdm6jHX9SjPja?dl=0


**Failed tests:**
* test_vm_life_cycle.py

 * test_10_attachAndDetach_iso Failing since 4 runs


**Skipped tests:**
test_vm_nic_adapter_vmxnet3
test_static_role_account_acls
test_deploy_vgpu_enabled_vm

**Passed test suits:**
test_deploy_vm_with_userdata.py
test_affinity_groups_projects.py
test_portable_publicip.py
test_vpc_vpn.py
test_over_provisioning.py
test_global_settings.py
test_scale_vm.py
test_service_offerings.py
test_routers_iptables_default_policy.py
test_routers.py
test_reset_vm_on_reboot.py
test_snapshots.py
test_deploy_vms_with_varied_deploymentplanners.py
test_login.py
test_list_ids_parameter.py
test_public_ip_range.py
test_multipleips_per_nic.py
test_regions.py
test_affinity_groups.py
test_network_acl.py
test_pvlan.py
test_volumes.py
test_nic.py
test_deploy_vm_root_resize.py
test_resource_detail.py
test_secondary_storage.py
test_disk_offerings.py

**Comment by ustcweizhou** [comment]
@bvbharatk I think a better idea for minimise the downtime in restartnetwork (with cleanup)/restartvpc (with cleanup)/ updatenetwork (to another redundant network offering) is 
(1) stop backup router (2) create new backup router (3) stop master (4) create a new router as backup.
as I posted in #1198 , I already implemented the features (restart network/vpc, update network) 4.7. I can create PRs if anyone is interested .

**Comment by remibergsma** [comment]
@ustcweizhou Would be great to see a PR for that 👍 thanks!

**Comment by bvbharatk** [comment]
@ustcweizhou any update regarding your changes.

**Comment by bvbharatk** [comment]
### ACS CI BVT Run
 **Sumarry:**
 Build Number 89
 Hypervisor xenserver
 NetworkType Advanced
 Passed=102
 Failed=1
 Skipped=4

_Link to logs Folder (search by build_no):_ https://www.dropbox.com/sh/yj3wnzbceo9uef2/AAB6u-Iap-xztdm6jHX9SjPja?dl=0


**Failed tests:**
* test_non_contigiousvlan.py

 * test_extendPhysicalNetworkVlan Failing since 9 runs


**Skipped tests:**
test_vm_nic_adapter_vmxnet3
test_static_role_account_acls
test_3d_gpu_support
test_deploy_vgpu_enabled_vm

**Passed test suits:**
test_deploy_vm_with_userdata.py
test_affinity_groups_projects.py
test_portable_publicip.py
test_over_provisioning.py
test_global_settings.py
test_scale_vm.py
test_service_offerings.py
test_routers_iptables_default_policy.py
test_loadbalance.py
test_routers.py
test_reset_vm_on_reboot.py
test_snapshots.py
test_deploy_vms_with_varied_deploymentplanners.py
test_network.py
test_router_dns.py
test_login.py
test_deploy_vm_iso.py
test_list_ids_parameter.py
test_public_ip_range.py
test_multipleips_per_nic.py
test_regions.py
test_affinity_groups.py
test_network_acl.py
test_pvlan.py
test_volumes.py
test_ssvm.py
test_nic.py
test_deploy_vm_root_resize.py
test_resource_detail.py
test_secondary_storage.py
test_vm_life_cycle.py
test_disk_offerings.py

**Comment by karuturi** [comment]
the failing test(test_non_contigiousvlan.py) is not related to the PR. 
The changes looks good and are ready to merge with required LGTMs and tests.
@bvbharat Can you move the sql to change to 490to491 upgrade file and make the PR against 4.7 branch? once thats done, I will do the merge.
Thanks

**Comment by abhinandanprateek** [comment]
@karuturi Do we know why test_non_contigiousvlan.py is failing ?

**Comment by bvbharatk** [comment]
@abhinandanprateek 
Hi Abhi, 
This test is failing because of test case clean up issues. Blow are the log messages from marvin 
Job failed: {jobprocstatus : 0, created : u'2016-09-12T07:06:55-0400', cmd : u'org.apache.cloudstack.api.command.admin.network.UpdatePhysicalNetworkCmd', userid : u'050d9020-78cd-11e6-a70e-00163e174325', jobstatus : 2, jobid : u'992cfb19-2b76-47f0-89ae-9d700c46b3bf', jobresultcode : 530, jobresulttype : u'object', jobresult : {errorcode : 431, errortext : u'physicalnetwork 200 has allocated vnets in the range 731-732'}, accountid : u'050d5cc2-78cd-11e6-a70e-00163e174325'}

**Comment by abhinandanprateek** [comment]
@bvbharatk thanks for the update if possible file a jira ticket for this issue.

 I just want to note that this has nothing to do with current PR which looks good.

**Pull request merged**